### PR TITLE
Add ETag header for static responses

### PR DESCRIPTION
**Pull request by redraw**
Related to https://github.com/simonw/datasette/issues/1645

This PR adds ETag headers for static responses. Adding cache-control headers was previously discussed but nothing done yet, so I've thought adding ETag headers would drastically reduce bandwidth meanwhile. Specially, when quickly deploying to Vercel or Fly, and you don't own a domain, required to add a CDN on front.

----
📚 Documentation preview 📚: https://datasette--2306.org.readthedocs.build/en/2306/

### CODE REVIEW

**Comment by redraw** [comment]
I've branched from latest main, but black found files to be reformatted.

```
Oh no! 💥 💔 💥
65 files would be reformatted, 32 files would be left unchanged.
```

**Comment by redraw** [comment]
Done, and tests passes. However, there's still a weird runtime error. Checking now.

```
Traceback (most recent call last):
  File "/home/agustin/tmp/datasette/datasette/app.py", line 1668, in route_path
    response = await view(request, send)
  File "/home/agustin/tmp/datasette/datasette/utils/asgi.py", line 337, in inner_static
    await asgi_send_file(
  File "/home/agustin/tmp/datasette/datasette/utils/asgi.py", line 305, in asgi_send_file
    await send(
  File "/home/agustin/tmp/datasette/env/lib/python3.10/site-packages/asgi_csrf.py", line 104, in wrapped_send
    await send(event)
  File "/home/agustin/tmp/datasette/env/lib/python3.10/site-packages/uvicorn/protocols/http/h11_impl.py", line 514, in send
    output = self.conn.send(event=h11.EndOfMessage())
  File "/home/agustin/tmp/datasette/env/lib/python3.10/site-packages/h11/_connection.py", line 512, in send
    data_list = self.send_with_data_passthrough(event)
  File "/home/agustin/tmp/datasette/env/lib/python3.10/site-packages/h11/_connection.py", line 545, in send_with_data_passthrough
    writer(event, data_list.append)
  File "/home/agustin/tmp/datasette/env/lib/python3.10/site-packages/h11/_writers.py", line 67, in __call__
    self.send_eom(event.headers, write)
  File "/home/agustin/tmp/datasette/env/lib/python3.10/site-packages/h11/_writers.py", line 96, in send_eom
    raise LocalProtocolError("Too little data for declared Content-Length")
h11._util.LocalProtocolError: Too little data for declared Content-Length

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/agustin/tmp/datasette/env/lib/python3.10/site-packages/uvicorn/protocols/http/h11_impl.py", line 408, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/home/agustin/tmp/datasette/env/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 69, in __call__
    return await self.app(scope, receive, send)
  File "/home/agustin/tmp/datasette/datasette/utils/asgi.py", line 477, in __call__
    return await self.asgi(scope, receive, send)
  File "/home/agustin/tmp/datasette/datasette/utils/asgi.py", line 186, in __call__
    await self.app(scope, receive, send)
  File "/home/agustin/tmp/datasette/env/lib/python3.10/site-packages/asgi_csrf.py", line 108, in app_wrapped_with_csrf
    await app(scope, receive, wrapped_send)
  File "/home/agustin/tmp/datasette/datasette/app.py", line 1625, in __call__
    return await self.route_path(scope, receive, send, path)
  File "/home/agustin/tmp/datasette/datasette/app.py", line 1686, in route_path
    return await self.handle_exception(request, send, exception)
  File "/home/agustin/tmp/datasette/datasette/app.py", line 1807, in handle_exception
    await response.asgi_send(send)
  File "/home/agustin/tmp/datasette/datasette/utils/asgi.py", line 365, in asgi_send
    await send(
  File "/home/agustin/tmp/datasette/env/lib/python3.10/site-packages/asgi_csrf.py", line 104, in wrapped_send
    await send(event)
  File "/home/agustin/tmp/datasette/env/lib/python3.10/site-packages/uvicorn/protocols/http/h11_impl.py", line 520, in send
    raise RuntimeError(msg % message_type)
RuntimeError: Unexpected ASGI message 'http.response.start' sent, after response already completed.
```

**Comment by redraw** [comment]
Might be related to how I'm forwarding headers.

**Comment by redraw** [comment]
Done! fixed.

**Comment by redraw** [comment]
image

File: tests/test_utils.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 Tests for various datasette helper functions.
 """
 
+from unittest import mock
```
Comment (simonw) [inline tests/test_utils.py:5]: I don't think the code uses this, I'll remove it.

**Comment by simonw** [comment]
This is clearly a good idea. Adding it directly to `asgi_static()` looks like the right place, too. I was initial concerned that this might interfere with how `/database.db` downloads work - the most performance-sensitive part of the code for this because the file might be 1GB+ and I wouldn't want to have to calculate MD5 against that every time - but those are not served by `asgi_static()` and in fact it turns out they have ETag support already:

https://github.com/simonw/datasette/blob/261fc8d875c43d5421a6ee1096992fd636c2c278/datasette/views/database.py#L327-L340

File: tests/test_utils.py (hunk 1 above)
Comment (redraw) [reply tests/test_utils.py]: Thx

**Comment by redraw** [comment]
Yes. I tried adding it inside `asgi_send_file()` first, then a related test failed and found out that same part where you added ETag headers for downloading the whole db. Then, I realized it was better to move it outside it and `asgi_static()` was safe.

**Comment by simonw** [comment]
I deployed this branch to my own Datasette Cloud instance - here's the before:

CleanShot 2024-03-17 at 12 02 42@2x

And after:

CleanShot 2024-03-17 at 12 14 20@2x

Huge improvement!

**Pull request merged**